src/body.cpp

```diff
@@ -902,6 +903,145 @@ void EarthAtm::SetAtmosphereHeight(double height){
   earth_with_atm_radius = radius + atm_height;
 }
 
+
+
+/*
+----------------------------------------------------------------------
+         EMITTINGEARTHATM CLASS DEFINITIONS
+----------------------------------------------------------------------
+*/
+
+// constructor without passed energy, used by nuSQUIDSAtm
+EmittingEarthAtm::EmittingEarthAtm():EmittingEarthAtm(getResourcePath()+"/atmos_prod/PROD_MODEL_MCEQ.dat")
+{} 
+
+// constructor with passed energy space
+EmittingEarthAtm::EmittingEarthAtm(marray<double, 1> ESpace):EmittingEarthAtm(getResourcePath()+"/atmos_prod/PROD_MODEL_MCEQ.dat")
+{ Set_EmissionEnergies(ESpace); }  
+
+
+EmittingEarthAtm::EmittingEarthAtm(std::string filepath):EarthAtm()
+{
+
+  marray<double,2> atm_prod_model = quickread(getResourcePath()+"atmos_prod/PROD_MODEL_MCEQ.dat");
+  arraysize = atm_prod_model.extent(0);
+
+  diff_enu_prod.resize(arraysize);
+  diff_munu_prod.resize(arraysize);
+  atm_coszen.resize(arraysize);
+  atm_heights.resize(arraysize);
+  atm_energy.resize(arraysize);
+
+  for (unsigned int i=0; i < arraysize;i++){
+    atm_coszen[i] = atm_prod_model[i][0];
+    atm_heights[i] = atm_prod_model[i][1];
+    atm_energy[i] = atm_prod_model[i][2];
+    diff_enu_prod[i] = atm_prod_model[i][3];
+    diff_munu_prod[i] = atm_prod_model[i][4];
+  }
+  
+  
+  unique_sort(atm_coszen, czen_min, czen_max, czen_number);
+  coszens.resize(0, atm_coszen.size());
+  for (unsigned int i=0; i < czen_number;i++){
+    coszens[i] = atm_coszen[i];
+  }
+  
+  unique_sort(atm_heights, height_min, height_max, height_number);
+  heights.resize(0,atm_heights.size());
+  for (unsigned int i=0; i < height_number; i++){
+    heights[i] = atm_heights[i];
+  }
+  
+  unique_sort(atm_energy, energy_min, energy_max, energy_number);
+  energies.resize(0,atm_energy.size());
+  for (unsigned int i=0; i < energy_number; i++){
+    energies[i] = atm_energy[i];
+  }
+  
+  marray<double, 3> nuEmatrix{czen_number, height_number, energy_number};
+  marray<double, 3> nuMumatrix{czen_number, height_number, energy_number};
+  
+
+  std::ifstream InputFile(filepath);
```

Comment:
We shouldn't read the file again; unless I misunderstand we already have all of this data in `atm_prod_model`, so at a minimum we should loop over that instead of repaeting all of the text parsing.